webpush/encrypt.go

```diff
@@ -241,11 +294,42 @@ func newInfo(infoType string, context []byte) []byte {
 	info = append(info, []byte("Content-Encoding: ")...)
 	info = append(info, []byte(infoType)...)
 	info = append(info, 0)
-	info = append(info, []byte("P-256")...)
-	info = append(info, context...)
+	// For aes128gcm ctx is not needed.
+	if len(context) > 0 {
+		info = append(info, []byte("P-256")...)
+		info = append(info, context...)
+	}
 	return info
 }
 
+// Returns a keying material. See sections of
+// https://tools.ietf.org/html/draft-ietf-webpush-encryption-07#section-3.3
+func newKeyInfo(subscriptionPublicKey, serverPublicKey []byte) []byte {
+	var info []byte
+	info = append(info, keyInfoPrefix...)
+	info = append(info, subscriptionPublicKey...)
+	info = append(info, serverPublicKey...)
+	return info
+}
+
+// Returns an 86 octet header. See section 2.1 of:
+// https://tools.ietf.org/html/draft-ietf-httpbis-encryption-encoding-06#section-2.1
+func appendHeader(salt, serverPublicKey, ciphertext []byte) []byte {
+	var result []byte
+	// A push service is not required to support more than 4096 octets of
+	// payload body so the record size can be at most 4096.
+	plen := uint32(4096)
+	plenbuf := make([]byte, 5)
+	binary.BigEndian.PutUint32(plenbuf, plen)
+	binary.BigEndian.PutUint16(plenbuf[3:], uint16(len(serverPublicKey)))
+	result = append(result, salt...)
+	result = append(result, plenbuf...)
+	result = append(result, serverPublicKey...)
```

Comment:
Which part?
Also see: https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#section-3.1 and https://tools.ietf.org/html/draft-ietf-webpush-encryption-08#appendix-A